gapir/cc/stack.h

```diff
@@ -26,457 +26,176 @@
-        return false;
-      }
-      *p = mValue.p;
-      return true;
-    }
+    void copyTo(void* address) { memcpy(address, (void*)&mData, mDataSize); }
 
-    bool getTo(void** p) const {
-      if (mType != BaseType::AbsolutePointer) {
-        return false;
-      }
-      *p = mValue.p;
-      return true;
+    Entry& operator=(const Entry& rhs) {
+      this->mDataSize = rhs.mDataSize;
+      memcpy((void*)&mData, (void*)&rhs.mData, maxEntrySize);
+      return *this;
     }
 
-    BaseValue getBaseValue() const { return mValue.bv; }
-
-    // Return a string describing the stack entry.
-    // The pointer returned is only valid until the next call to debugInfo,
-    // regardless of the Entry instance. This function is not thread safe.
-    const char* debugInfo(const MemoryManager* memoryManager) const;
-
    private:
-    // Type of the element stored by this entry
-    BaseType mType;
-
-    // Union of all possible types stored on the stack for creating a unified
-    // value type with getter function to access the value as a specific type
-    union ValueType {
-      bool b;
-      int8_t i8;
-      int16_t i16;
-      int32_t i32;
-      int64_t i64;
-      uint8_t u8;
-      uint16_t u16;
-      uint32_t u32;
-      uint64_t u64;
-      float f;
-      double d;
-      void* p;
-      BaseValue bv;
-    };
-
-    ValueType mValue;
-
-    static_assert(sizeof(BaseValue) >= sizeof(ValueType),
-                  "Stack::BaseValue is not large enough");
+    constexpr static const size_t maxEntrySize = 8;
+    constexpr static const size_t entryAlignment = maxEntrySize;
+    typename std::aligned_storage<maxEntrySize, entryAlignment>::type mData;
+
+    size_t mDataSize;
+    std::type_index mDataType;
+    bool mIsVolatilePointer;
   };
 
-  // Indicates if the stack is in a consistent state (true value) or not (false
-  // value). The stack go into an inconsistent state after an invalid operation.
-  // When mValid is false then all of the operation on the stack (expect
-  // printing the stack) produce a warning message and falls back to no op (with
-  // zero initialized return value where necessary). The stack can't go back
-  // from an invalid state to a valid state again.
-  bool mValid;
-
-  // Contains the offset of the first empty slot in the stack from the bottom of
-  // the stack. The value of it indicates the number of elements currently in
-  // the stack.
-  uint32_t mTop;
-
-  // mStack stores the entries currently in the stack. The 0th element
-  // corresponds to the bottom of the stack.
-  std::vector<Entry> mStack;
-
-  // Reference to the memory manager used to resolve constant and volatile
-  // pointers to absolute pointers when thy are popped from the stack.
+  uint32_t mNumEntries;
+  uint32_t mMaxEntries;
+  Entry* mStack;
+
   const MemoryManager* mMemoryManager;
 };
 
+template <class T>
+void Stack::Entry::set(const T& val) {
+  assertCanStore<T>();
+  *((T*)&mData) = val;
```

Comment:
I think this is still not spec-valid, typically for this case you would memcpy in and out, rather than try to assign.

Reply:
Do you still object here since the change to std::aligned_storage?

Reply:
memcpy(&mData, &val, sizeof(val))

Would remain spec-valid, and prevent this issue. (Also fix your failing builds)